code/modules/reagents/chemistry/machinery/chem_heater.dm

```diff
@@ -131,18 +136,33 @@
 	data["beakerContents"] = beaker_contents
 
 	var/list/active_reactions = list()
+	var/flashing = 14 //for use with alertAfter - since there is no alertBefore, I set the after to 0 if true, or to the max value if false
 	for(var/_reaction in beaker?.reagents.reaction_list)
 		var/datum/equilibrium/equilibrium = _reaction
 		if(!equilibrium.reaction.results)//Incase of no result reactions
 			continue
 		active_reactions.len++
 		var/_reagent = equilibrium.reaction.results[1]
-		var/datum/reagent/reagent = equilibrium.holder.get_reagent(_reagent) //Reactions are named after their primary products
-		active_reactions[length(active_reactions)] = list("name" = reagent.name, "purity" = round(reagent.purity, 0.01), "overheat" = equilibrium.reaction.overheat_temp, "barColor" = get_purity_color(equilibrium), "reactedVol" = equilibrium.reacted_vol, "targetVol" = round(equilibrium.target_vol, 1))//Use the first result reagent to name the reaction detected
+		var/datum/reagent/reagent = beaker?.reagents.get_reagent(_reagent) //Reactions are named after their primary products
+		//if(!reagent)
+		//	continue
+		var/danger = FALSE
+		var/purity_alert = 2 //same as flashing
+		if(reagent.purity < equilibrium.reaction.purity_min)
+			purity_alert = ENABLE_FLASHING//Because 0 is seen as null
+			danger = TRUE
+		if((!flashing == ENABLE_FLASHING) && (upgrade_level > 1))//So that the pH meter flashes for ANY reactions out of optimal
```

Comment:
`(!flashing == ENABLE_FLASHING)`

This'll never pass. Did you mean `!(flashing == ENABLE_FLASHING)`?

Reply:
That's why I was having so much trouble with this! Thank you.